tests/unit/azure/test_credentials.py

```diff
@@ -0,0 +1,337 @@
+from databricks.sdk import WorkspaceClient
+from databricks.sdk.errors import NotFound, ResourceDoesNotExist
+from databricks.sdk.errors.platform import InvalidParameterValue
+from databricks.sdk.service.catalog import (
+    AwsIamRole,
+    AzureManagedIdentity,
+    AzureServicePrincipal,
+    StorageCredentialInfo,
+    ValidateStorageCredentialResponse,
+)
+from databricks.sdk.service.workspace import GetSecretResponse
+
+from databricks.labs.ucx.assessment.azure import (
+    AzureServicePrincipalCrawler,
+    AzureServicePrincipalInfo,
+)
+from databricks.labs.ucx.azure.access import (
+    AzureResourcePermissions,
+    StoragePermissionMapping,
+)
+from databricks.labs.ucx.azure.credentials import (
+    ServicePrincipalMigration,
+    ServicePrincipalMigrationInfo,
+    StorageCredentialManager,
+)
+
+
+@pytest.fixture
+def ws():
+    state = {"/Users/foo/.ucx/config.yml": yaml.dump({'version': 2, 'inventory_database': 'ucx'})}
+
+    def download(path: str) -> io.StringIO:
+        if path not in state:
+            raise NotFound(path)
+        return io.StringIO(state[path])
+
+    ws_mock = create_autospec(WorkspaceClient)
+    ws_mock.config.host = 'https://localhost'
+    ws_mock.current_user.me().user_name = "foo"
+    ws_mock.workspace.download = download
+    return ws_mock
+
+
+def side_effect_create_storage_credential(name, azure_service_principal, comment, read_only):
+    return StorageCredentialInfo(
+        name=name, azure_service_principal=azure_service_principal, comment=comment, read_only=read_only
+    )
+
+
+def side_effect_validate_storage_credential(storage_credential_name, url, read_only):  # pylint: disable=unused-argument
+    if "overlap" in storage_credential_name:
+        raise InvalidParameterValue
+    if read_only:
+        response = {"isDir": True, "results": [{"message": "", "operation": "READ", "result": "PASS"}]}
+        return ValidateStorageCredentialResponse.from_dict(response)
+    response = {"isDir": True, "results": [{"message": "", "operation": "WRITE", "result": "PASS"}]}
+    return ValidateStorageCredentialResponse.from_dict(response)
+
+
+@pytest.fixture
+def credential_manager(ws):
+    ws.storage_credentials.list.return_value = [
+        StorageCredentialInfo(aws_iam_role=AwsIamRole("arn:aws:iam::123456789012:role/example-role-name")),
+        StorageCredentialInfo(
+            azure_managed_identity=AzureManagedIdentity("/subscriptions/.../providers/Microsoft.Databricks/...")
+        ),
+        StorageCredentialInfo(
+            azure_service_principal=AzureServicePrincipal(
+                "62e43d7d-df53-4c64-86ed-c2c1a3ac60c3",
+                "b6420590-5e1c-4426-8950-a94cbe9b6115",
+                "secret",
+            )
+        ),
+        StorageCredentialInfo(azure_service_principal=AzureServicePrincipal("directory_id_1", "app_secret2", "secret")),
+    ]
+
+    ws.storage_credentials.create.side_effect = side_effect_create_storage_credential
```

Comment:
```suggestion
    ws.storage_credentials.create.return_value = StorageCredentialInfo(
        name=name, azure_service_principal=azure_service_principal, comment=comment, read_only=read_only
    )
```
don't make code more complicated than it should be